lib/Sema/CSSimplify.cpp

```diff
@@ -4960,6 +4960,29 @@ ConstraintSystem::addKeyPathApplicationConstraint(Type keypath,
                                               Type root, Type value,
                                               ConstraintLocatorBuilder locator,
                                               bool isFavored) {
+  // If this is a subscript with a KeyPath expression, add a constraint that
+  // connects the subscript's root type to the root type of the KeyPath.
+  SmallVector<LocatorPathElt, 4> path;
+  Expr *anchor = locator.getLocatorParts(path);
+  if (auto subscript = dyn_cast_or_null<SubscriptExpr>(anchor)) {
+    if (path.size() == 1 &&
+        path[0].getKind() == ConstraintLocator::SubscriptMember) {
+      if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
+        if (indexTuple->getNumElements() == 1) {
+          if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
```

Comment:
I suspect some of the above logic could be turned into `cast<>()` and `assert`s based on our expectations of what we know when we decide an overload is a key path application.

For whatever remains, it might be nice to remove some of the depth of control flow via the advice in https://www.llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

You could for example make this a closure with early returns, a separate function, or a:
```
  do {
    // use break in places where we bail out
  } while (0);